src/librustc_codegen_ssa/mir/rvalue.rs

```diff
@@ -293,12 +293,17 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                                 if let Some(discr) =
                                     operand.layout.ty.discriminant_for_variant(bx.tcx(), index)
                                 {
-                                    let discr_val = bx.cx().const_uint_big(ll_t_out, discr.val);
+                                    let discr_layout = bx.cx().layout_of(discr.ty);
+                                    let discr_t = bx.cx().immediate_backend_type(discr_layout);
+                                    let discr_val = bx.cx().const_uint_big(discr_t, discr.val);
+                                    let discr_val =
+                                        bx.intcast(discr_val, ll_t_out, discr.ty.is_signed());
+
                                     return (
                                         bx,
                                         OperandRef {
                                             val: OperandValue::Immediate(discr_val),
-                                            layout: cast,
+                                            layout: discr_layout,
```

Comment:
Is this correct? After all, you have cast `discr_val` to `ll_t_out` already, so why would we still use the `discr_layout` layout here?

Reply:
No, I think you're right.

Reply:
@eddyb please check this (see above)

Reply:
Look correct on the basis of Int -> Int casts (implemented below).